README.md
# feofetch

Yet another (x)fetch tool, written in rust with cross-platform in mind.

## Note on cross-platform

This program was written specifically with cross-platform libraries, so in
theory, this should compile and run on Windows, MacOS, and Linux. However, this
program so far has only been tested on Linux, and a lot of functionality will be
missing on other OSes.

Any feature requests, bug reports, or pull requests regarding this are very
welcome!

## Installation

### With cargo

`feofetch` can be installed with
[cargo](https:#www.rust-lang.org/tools/install).

```sh
cargo install feofetch
```

### Build from source

If you want the latest build, you can build directly from source. This requires
[rust](https:#www.rust-lang.org/tools/install) to be installed on your system.

```sh
git clone https:#github.com/ProspectPyxis/feofetch.git
# Or download the source code zip file directly and extract

cd feofetch/
cargo install --path .
```

## Usage

Just run `feofetch` in the terminal:

```sh
feofetch
```

The program also supports various command line flags.

```
Usage: feofetch [OPTIONS]

Options:
  -c, --config-path <CONFIG_PATH>      Override default config location with the specified path
  -x, --x-offset <X_OFFSET>            Set x-offset to this number, overriding config
  -y, --y-offset <Y_OFFSET>            Set y-offset to this number, overriding config
  -p, --padding-lines <PADDING_LINES>  Set padding lines to this number, overriding config
  -h, --help                           Print help information
  -V, --version                        Print version information
```

## Configuration

For how to configure this program, see [CONFIGURATION.md](CONFIGURATION.md).

## Contributing

Any pull requests, issues, or feature requests are welcome!

## License

This project is licensed under either of
[Apache License, Version 2.0](LICENSE-APACHE.md) or
[MIT License](LICENSE-MIT.md), at your option.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Program now respects the `NO_COLOR` environment variable, thanks to porting
  from `crossterm` to `termcolor`

### Changed

- ASCII text that has less lines than the amount of data types is now centered
  vertically

## [0.2.0] - 2022-12-18

### Breaking Changes

- Restructured config file - split options specific to data types to its own
  tables.

### Added

- Command line flags: `--help`, `--version`.
- `--config-path`/`-c` command line flag to set an alternate config path for one
  execution.
- `--x-offset`/`-x` and `--y-offset`/`-y` command line flags fo override offsets
  for this run.
- `--padding-lines`/`-p` command line flag to override padding lines for this
  run.
- Highly improved error handling - program should no longer panic other than in
  catastrophic cases, thanks to implementing the `anyhow` crate.
- New data options: `shell`, `terminal`.
- `label_color` and `ascii.color` config options that supports all ANSI colors.

### Changed

- Better icons for `os` data type - now shows the appropriate logo for Windows,
  Mac, and Linux for its icons.

## [0.1.0] - 2022-07-03

### Added

- Data options: `os`, `version`, `uptime`, `packages`, `wm`.
- `use_icons` config option.
- Ability to display ASCII art alongside data.


src/data.rs
use crate::{
	config::{self, Config, FetchType},
	packages,
};
use crossterm::{
	queue,
	style::{Print, PrintStyledContent, Stylize},
};
use std::{
	env,
	io::{self, Stdout},
};
use which::which;

pub struct FetchData {
	pub label: &'static str,
	pub icon: &'static str,
	text: String,
}

impl FetchData {
	pub fn get(data: &FetchType, conf: &Config) -> Self {
		match data {
			FetchType::Os => FetchData {
				label: "os",
				icon: "",
				text: config::get_os(),
			},

			FetchType::Version => FetchData {
				label: "version",
				icon: "",
				text: {
					let mut version =
						sys_info::os_release().unwrap_or_else(|_| "unknown".to_string());
					if let Some(index) = version.find('-') {
						version.truncate(index);
					}
					version
				},
			},

			FetchType::Uptime => FetchData {
				label: "uptime",
				icon: "",
				text: match uptime_lib::get() {
					Ok(time) => {
						let time = time.as_secs();
						let mut timestr = Vec::new();
						if time >= 3600 {
							timestr.push(format!("{}h", time / 3600));
						}
						if time % 3600 > 0 {
							timestr.push(format!("{}m", (time % 3600) / 60));
						}
						if !timestr.is_empty() {
							timestr.join(" ")
						} else {
							"0m".to_string()
						}
					}
					Err(_) => "unknown".to_string(),
				},
			},

			FetchType::Packages => FetchData {
				label: "packages",
				icon: "",
				text: packages::get_packages(conf.packages.print_package_manager_names),
			},

			FetchType::Wm => FetchData {
				label: "wm",
				icon: "",
				text: {
					let get_no_wmctrl = || match sys_info::os_type()
						.unwrap_or_else(|_| "Unknown".to_string())
						.as_str()
					{
						"Linux" => env::var("XDG_SESSION_DESKTOP")
							.or_else(|_| env::var("DESKTOP_SESSION"))
							.unwrap_or_else(|_| "unknown".to_string()),
						// TODO: Handle windows/mac os
						_ => "unknown".to_string(),
					};

					if conf.wm.use_wmctrl && which("wmctrl").is_ok() {
						let try_get_wmctrl = || -> Result<String, ()> {
							let out = std::process::Command::new("wmctrl")
								.arg("-m")
								.output()
								.map_err(|_| ())?;

							let s = std::str::from_utf8(&out.stdout).map_err(|_| ())?;

							match s.lines().find(|x| x.starts_with("Name: ")) {
								Some(line) => {
									Ok(line.strip_prefix("Name: ").unwrap_or("error").to_string())
								}
								None => Ok("unknown".to_string()),
							}
						};

						match try_get_wmctrl() {
							Ok(wm) => wm,
							Err(_) => get_no_wmctrl(),
						}
					} else {
						get_no_wmctrl()
					}
				},
			},
		}
	}

	pub fn queue_print(
		&self,
		stdout: &mut Stdout,
		data_pos: usize,
		conf: &Config,
	) -> Result<(), io::Error> {
		let label_text = if conf.use_icons {
			self.icon
		} else {
			self.label
		};

		queue!(
			stdout,
			PrintStyledContent(format!("{:data_pos$}", label_text).bold().cyan()),
			Print(&self.text),
		)
	}
}

pub fn fetch_all(conf: &Config) -> Vec<FetchData> {
	conf.data.iter().map(|d| FetchData::get(d, conf)).collect()
}


src/main.rs
mod args;
mod config;
mod fetch;
mod packages;

use anyhow::Context;
use clap::Parser;
use etcetera::app_strategy::{self, AppStrategy, AppStrategyArgs};
use std::fs::{canonicalize, read_to_string};

fn main() -> anyhow::Result<()> {
	let strategy = app_strategy::choose_app_strategy(AppStrategyArgs {
		top_level_domain: "com".to_string(),
		author: "ProspectPyxis".to_string(),
		app_name: "feofetch".to_string(),
	})
	.context("Failed to set up app strategy")?;

	let args = args::Args::parse();
	let config_path = match args.config_path {
		Some(ref path) => {
			if path.eq_ignore_ascii_case("default") {
				strategy.in_config_dir("config.toml")
			} else {
				canonicalize(path)
					.with_context(|| format!("Failed to parse config TOML path {}", path))?
			}
		}
		None => strategy.in_config_dir("config.toml"),
	};
	let conf = config::get_config(&config_path)
		.with_context(|| {
			format!(
				"Failed to read config file {}",
				config_path.to_string_lossy(),
			)
		})?
		.with_overrides(&args);
	let ascii_file = match conf.ascii.ascii_path {
		Some(ref path) => path
			.parse()
			.with_context(|| format!("Failed to parse ascii path {}", path))?,
		None => strategy.in_config_dir("ascii.txt"),
	};
	let ascii = conf
		.ascii
		.print
		.then_some(read_to_string(&ascii_file).with_context(|| {
			format!(
				"Failed to read ascii file from {}",
				ascii_file.to_string_lossy()
			)
		})?);

	let data = fetch::fetch_all(&conf);
	fetch::print_all_fetches(&data, &conf, ascii.as_deref())
		.context("Failed to write to stdout")?;

	Ok(())
}


src/config.rs
use crate::args::Args;
use anyhow::Context;
use serde::Deserialize;
use std::{default::Default, fs, io::ErrorKind, path::PathBuf};
use termcolor::Color;

#[derive(Deserialize)]
#[serde(rename_all = "snake_case")]
#[non_exhaustive]
pub enum FetchType {
	Os,
	Version,
	Uptime,
	Packages,
	Wm,
	Shell,
	Terminal,
}

#[derive(Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ColorConfig {
	Normal,
	Black,
	#[serde(alias = "dark_gray")]
	DarkGrey,
	Red,
	DarkRed,
	Green,
	DarkGreen,
	Yellow,
	DarkYellow,
	Blue,
	DarkBlue,
	Magenta,
	DarkMagenta,
	Cyan,
	DarkCyan,
	White,
	#[serde(alias = "gray")]
	Grey,
}

#[derive(Deserialize)]
#[serde(default)]
pub struct Config {
	pub data: Vec<FetchType>,
	pub use_icons: bool,
	pub label_color: ColorConfig,
	pub align_spaces: usize,
	pub offset: (usize, usize),
	pub padding_lines: usize,
	pub ascii: AsciiConfig,
	pub packages: PackagesConfig,
	pub wm: WmConfig,
}

#[derive(Deserialize)]
#[serde(default)]
pub struct AsciiConfig {
	pub print: bool,
	pub ascii_path: Option<String>,
	pub color: ColorConfig,
	pub align_spaces: usize,
}

#[derive(Default, Deserialize)]
#[serde(default)]
pub struct PackagesConfig {
	pub print_package_manager_names: bool,
}

#[derive(Default, Deserialize)]
#[serde(default)]
pub struct WmConfig {
	pub use_wmctrl: bool,
}

impl ColorConfig {
	pub fn get_color(&self) -> Option<Color> {
		match self {
			ColorConfig::Normal => None,
			ColorConfig::Black | ColorConfig::DarkGrey => Some(Color::Black),
			ColorConfig::Red | ColorConfig::DarkRed => Some(Color::Red),
			ColorConfig::Green | ColorConfig::DarkGreen => Some(Color::Green),
			ColorConfig::Yellow | ColorConfig::DarkYellow => Some(Color::Yellow),
			ColorConfig::Blue | ColorConfig::DarkBlue => Some(Color::Blue),
			ColorConfig::Magenta | ColorConfig::DarkMagenta => Some(Color::Magenta),
			ColorConfig::Cyan | ColorConfig::DarkCyan => Some(Color::Cyan),
			ColorConfig::White | ColorConfig::Grey => Some(Color::White),
		}
	}

	pub fn is_intense(&self) -> bool {
		let bright_colors = [
			ColorConfig::DarkGrey,
			ColorConfig::Red,
			ColorConfig::Green,
			ColorConfig::Yellow,
			ColorConfig::Blue,
			ColorConfig::Magenta,
			ColorConfig::Cyan,
			ColorConfig::White,
		];

		bright_colors.contains(self)
	}
}

impl Default for Config {
	fn default() -> Self {
		Config {
			data: vec![
				FetchType::Os,
				FetchType::Version,
				FetchType::Uptime,
				FetchType::Packages,
				FetchType::Wm,
			],
			use_icons: false,
			label_color: ColorConfig::Cyan,
			align_spaces: 2,
			offset: (0, 0),
			padding_lines: 1,
			ascii: AsciiConfig::default(),
			packages: PackagesConfig::default(),
			wm: WmConfig::default(),
		}
	}
}

impl Config {
	pub fn with_overrides(self, args: &Args) -> Self {
		Config {
			offset: (
				args.x_offset.unwrap_or(self.offset.0),
				args.y_offset.unwrap_or(self.offset.1),
			),
			padding_lines: args.padding_lines.unwrap_or(self.padding_lines),
			..self
		}
	}
}

impl Default for AsciiConfig {
	fn default() -> Self {
		AsciiConfig {
			print: false,
			ascii_path: None,
			color: ColorConfig::Normal,
			align_spaces: 2,
		}
	}
}

pub fn get_config(config_path: &PathBuf) -> anyhow::Result<Config> {
	let config = match fs::read_to_string(config_path) {
		Ok(config_str) => toml::from_str(&config_str).with_context(|| {
			format!(
				"Failed to parse config TOML at {}",
				config_path.to_string_lossy()
			)
		})?,
		Err(e) => match e.kind() {
			ErrorKind::NotFound => Config::default(),
			_ => {
				return Err(e).with_context(|| {
					format!(
						"Failed to read config TOML from {}",
						config_path.to_string_lossy()
					)
				})
			}
		},
	};
	Ok(config)
}


src/packages.rs
use anyhow::Context;
use which::which;

struct PacManData {
	check_name: &'static str,
	display_name: &'static str,
	cmd_name: &'static str,
	args: &'static [&'static str],
}

// TODO: Add more package managers
const PACKAGE_MANAGERS: &[PacManData] = &[
	PacManData {
		check_name: "pacman-key",
		display_name: "pacman",
		cmd_name: "pacman",
		args: &["-Qq"],
	},
	PacManData {
		check_name: "dpkg",
		display_name: "dpkg",
		cmd_name: "dpkg-query",
		args: &["-f", ".\n", "-W"],
	},
	PacManData {
		check_name: "brew",
		display_name: "homebrew",
		cmd_name: "brew",
		args: &["list", "-1"],
	},
];

fn try_get_package_count(pacman: &PacManData) -> anyhow::Result<usize> {
	let out = std::process::Command::new(pacman.cmd_name)
		.args(pacman.args)
		.output()
		.with_context(|| {
			format!(
				"Failed to get stdout while running command {}",
				pacman.cmd_name
			)
		})?;

	let packages_count = std::str::from_utf8(&out.stdout).with_context(|| {
		format!(
			"Failed to parse stdout of command {} to string",
			pacman.cmd_name
		)
	})?;

	Ok(packages_count.lines().count())
}

pub fn get_packages(display_package_manager: bool) -> String {
	let packages_count = PACKAGE_MANAGERS
		.iter()
		.filter(|pacman| which(pacman.check_name).is_ok())
		.fold(0, |accum, pacman| {
			accum
				+ try_get_package_count(pacman).unwrap_or_else(|e| {
					eprintln!("{:#}", e);
					eprintln!("Assuming 0 packages installed by this package manager, moving on");
					0
				})
		});
	let pacmans: Vec<&'static str> = PACKAGE_MANAGERS
		.iter()
		.filter(|pacman| which(pacman.check_name).is_ok())
		.map(|pacman| pacman.display_name)
		.collect();

	if pacmans.is_empty() {
		"unknown".to_string()
	} else if display_package_manager {
		format!("{} ({})", packages_count, pacmans.join(", "))
	} else {
		format!("{}", packages_count)
	}
}


src/fetch.rs
use crate::{
	config::{Config, FetchType},
	packages,
};
use std::{
	env,
	io::{self, Write},
};
use termcolor::{ColorChoice, ColorSpec, StandardStream, WriteColor};

mod wm {
	use crate::config::Config;
	use std::env;
	use which::which;

	fn get_no_wmctrl() -> String {
		match &sys_info::os_type().unwrap_or_else(|_| "Unknown".to_string())[..] {
			"Linux" => env::var("XDG_SESSION_DESKTOP")
				.or_else(|_| env::var("DESKTOP_SESSION"))
				.unwrap_or_else(|_| "unknown".to_string()),
			// TODO: Handle windows/mac os
			_ => "unknown".to_string(),
		}
	}

	fn try_get_wmctrl() -> Option<String> {
		let out = std::process::Command::new("wmctrl")
			.arg("-m")
			.output()
			.ok()?;

		let output_str = std::str::from_utf8(&out.stdout).ok()?;

		Some(
			output_str
				.lines()
				.find(|line| line.starts_with("Name: "))
				.map(|line| line.strip_prefix("Name: ").unwrap_or("error").to_string())
				.unwrap_or_else(|| "unknown".to_string()),
		)
	}

	pub fn get(conf: &Config) -> String {
		if conf.wm.use_wmctrl && which("wmctrl").is_ok() {
			try_get_wmctrl().unwrap_or_else(get_no_wmctrl)
		} else {
			get_no_wmctrl()
		}
	}
}

pub struct FetchData {
	pub label: &'static str,
	pub icon: &'static str,
	text: String,
}

impl FetchData {
	pub fn get(data: &FetchType, conf: &Config) -> Self {
		match data {
			FetchType::Os => {
				let os = sys_info::os_type().unwrap_or_else(|_| "Unknown".to_string());
				let os_str = if &os[..] == "Linux" {
					sys_info::linux_os_release()
						.ok()
						.and_then(|info| info.id)
						.unwrap_or_else(|| "linux".to_string())
				} else {
					os.to_lowercase()
				};

				FetchData {
					label: "os",
					icon: match &os[..] {
						"Linux" => "",
						"Darwin" => "",
						"Windows" => "",
						_ => "",
					},
					text: os_str,
				}
			}

			FetchType::Version => FetchData {
				label: "version",
				icon: "",
				text: {
					let mut version =
						sys_info::os_release().unwrap_or_else(|_| "unknown".to_string());
					if let Some(index) = version.find('-') {
						version.truncate(index);
					}
					version
				},
			},

			FetchType::Uptime => FetchData {
				label: "uptime",
				icon: "",
				text: match uptime_lib::get() {
					Ok(time) => {
						let time = time.as_secs();
						let mut timestr = Vec::new();
						if time >= 3600 {
							timestr.push(format!("{}h", time / 3600));
						}
						if time % 3600 > 0 {
							timestr.push(format!("{}m", (time % 3600) / 60));
						}
						if !timestr.is_empty() {
							timestr.join(" ")
						} else {
							"0m".to_string()
						}
					}
					Err(_) => "unknown".to_string(),
				},
			},

			FetchType::Packages => FetchData {
				label: "packages",
				icon: "",
				text: packages::get_packages(conf.packages.print_package_manager_names),
			},

			FetchType::Wm => FetchData {
				label: "wm",
				icon: "",
				text: wm::get(conf),
			},

			FetchType::Shell => FetchData {
				label: "shell",
				icon: "",
				text: {
					env::var("SHELL")
						.ok()
						.and_then(|shell| {
							std::path::PathBuf::from(&shell)
								.file_name()
								.map(|s| s.to_string_lossy().to_string())
						})
						.unwrap_or_else(|| "unknown".to_string())
				},
			},

			FetchType::Terminal => FetchData {
				label: "terminal",
				icon: "",
				text: {
					match env::var("TERM_PROGRAM") {
						Ok(term) => term
							.strip_suffix(".app")
							.map(|term| term.to_string())
							.unwrap_or(term),
						Err(_) => "unknown".to_string(),
					}
				},
			},
		}
	}

	pub fn queue_print(
		&self,
		stdout: &mut StandardStream,
		data_pos: usize,
		conf: &Config,
	) -> Result<(), io::Error> {
		stdout.set_color(
			ColorSpec::new()
				.set_fg(conf.label_color.get_color())
				.set_intense(conf.label_color.is_intense())
				.set_bold(true),
		)?;
		write!(
			stdout,
			"{:data_pos$}",
			if conf.use_icons {
				self.icon
			} else {
				self.label
			},
		)?;
		stdout.set_color(&ColorSpec::new())?;
		write!(stdout, "{}", self.text)?;
		Ok(())
	}
}

pub fn fetch_all(conf: &Config) -> Vec<FetchData> {
	conf.data.iter().map(|d| FetchData::get(d, conf)).collect()
}

pub fn print_all_fetches(
	data: &[FetchData],
	conf: &Config,
	ascii: Option<&str>,
) -> Result<(), io::Error> {
	let (ascii_lines_count, ascii_max_length) = match ascii {
		Some(a) => (
			a.lines().count(),
			a.lines().fold(0, |acc, x| acc.max(x.chars().count())) + conf.ascii.align_spaces,
		),
		None => (0, 0),
	};
	let data_start = ascii_lines_count.saturating_sub(data.len()) / 2;
	let ascii_start = data.len().saturating_sub(ascii_lines_count) / 2;
	let total_lines = ascii_lines_count.max(data.len());

	let data_pos = if !conf.use_icons {
		data.iter()
			.fold(0, |acc, x| acc.max(x.label.chars().count()))
	} else {
		1
	} + conf.align_spaces;

	let mut stdout = StandardStream::stdout(ColorChoice::Auto);

	write!(&mut stdout, "{}", "\n".repeat(conf.offset.1))?;

	let mut ascii_lines = ascii.unwrap_or("").lines();
	let mut data_lines = data.iter();
	for index in 0..total_lines {
		write!(&mut stdout, "{}", " ".repeat(conf.offset.0))?;

		if index >= ascii_start {
			match ascii_lines.next() {
				Some(line) => {
					stdout.set_color(
						ColorSpec::new()
							.set_fg(conf.ascii.color.get_color())
							.set_intense(conf.ascii.color.is_intense())
							.set_bold(true),
					)?;
					write!(&mut stdout, "{:ascii_max_length$}", line)
				}
				None => write!(&mut stdout, "{}", " ".repeat(ascii_max_length)),
			}
		} else {
			write!(&mut stdout, "{}", " ".repeat(ascii_max_length))
		}?;

		if index >= data_start {
			if let Some(line) = data_lines.next() {
				line.queue_print(&mut stdout, data_pos, conf)?;
			}
		}

		writeln!(&mut stdout)?;
	}

	write!(&mut stdout, "{}", "\n".repeat(conf.padding_lines))?;

	Ok(())
}
